### PR TITLE
fix: handle untyped constant expressions in append()

### LINKED ISSUE #1136: can't append `nil` to a slice of interfaces
```go
package main

func main() {
        var errors []error
        errors = append(errors, nil)
        _ = errors
}
```

```console
$ gno run a.gno
--- preprocess stack ---
stack 2: func main() { var errors<VPBlock(1,0)> [](const-type .uverse.error); errors<VPBlock(1,0)> = (const (append func(x <X>{},args MaybeNativeType{...<X.Elem()>{}})(res <X>{})))(error
s<VPBlock(1,0)>, (const (undefined))); _<VPUverse(0)> = errors<VPUverse(0)> }
stack 1: file{ package main; func main() { var errors<VPBlock(1,0)> [](const-type .uverse.error); errors<VPBlock(1,0)> = (const (append func(x <X>{},args MaybeNativeType{...<X.Elem()>{}}
)(res <X>{})))(errors<VPBlock(1,0)>, (const (undefined))); _<VPUverse(0)> = errors<VPUverse(0)> } }
stack 0: package(main)
------------------------
panic: unspecified generic varg args MaybeNativeType{...<X.Elem()>{}} [recovered]
        panic: main/a.gno:5: unspecified generic varg args MaybeNativeType{...<X.Elem()>{}}

goroutine 1 [running]:
github.com/gnolang/gno/gnovm/pkg/gnolang.Preprocess.func2.1()
        /home/howl/oc/gno/gnovm/pkg/gnolang/preprocess.go:173 +0x46d
panic({0xbf0b20, 0xc00007d560})
        /usr/lib/go/src/runtime/panic.go:890 +0x263
github.com/gnolang/gno/gnovm/pkg/gnolang.(*FuncType).Specify(0xc0000ac5f0, {0xf651b0, 0xc0001881b0}, {0xc0001b79a0, 0x2, 0x2}, 0x0)
        /home/howl/oc/gno/gnovm/pkg/gnolang/types.go:1174 +0x1a54
github.com/gnolang/gno/gnovm/pkg/gnolang.Preprocess.func2({0xc000212a00?, 0x3, 0x20}, 0x24, 0x40ae04?, {0xf5acf8, 0xc0002fd620?}, 0x3)
        /home/howl/oc/gno/gnovm/pkg/gnolang/preprocess.go:1096 +0x3c6e
github.com/gnolang/gno/gnovm/pkg/gnolang.transcribe(0xc0001a9970, {0xc000212a00?, 0x3, 0x20}, 0x24, 0xf5aaf0?, {0xf5acf8?, 0xc0002fd620}, 0xc0001a9217)
        /home/howl/oc/gno/gnovm/pkg/gnolang/transcribe.go:728 +0x4b63
github.com/gnolang/gno/gnovm/pkg/gnolang.transcribe(0xc0001a9970, {0xc000212a00?, 0x2, 0x20}, 0x46, 0xf5b170?, {0xf5aaf0?, 0xc0002fd680}, 0xc0001a9447)
        /home/howl/oc/gno/gnovm/pkg/gnolang/transcribe.go:371 +0x6e99
```

### LINKED ISSUE #1149: in `append([]T, ...T)`, constant types are not converted to T implicitly
```go
package main

func main() {
        const x = 118999
        y := 11
        p := []int{}
        p = append(p, x, y)
}
```

```console
$ gno run a.gno
--- preprocess stack ---
stack 2: func main() { const x<VPBlock(1,0)> = (const (118999 <untyped> bigint)); y<VPBlock(1,1)> := (const (11 int)); p<VPBlock(1,2)> := [](const-type int){}; p<VPBlock(1,2)> = (const (
append func(x <X>{},args MaybeNativeType{...<X.Elem()>{}})(res <X>{})))(p<VPBlock(1,2)>, (const (118999 <untyped> bigint)), y<VPBlock(1,1)>) }
stack 1: file{ package main; func main() { const x<VPBlock(1,0)> = (const (118999 <untyped> bigint)); y<VPBlock(1,1)> := (const (11 int)); p<VPBlock(1,2)> := [](const-type int){}; p<VPBl
ock(1,2)> = (const (append func(x <X>{},args MaybeNativeType{...<X.Elem()>{}})(res <X>{})))(p<VPBlock(1,2)>, (const (118999 <untyped> bigint)), y<VPBlock(1,1)>) } }
stack 0: package(main)
------------------------
panic: incompatible varg types: expected <untyped> bigint, got int [recovered]
        panic: main/a.gno:7: incompatible varg types: expected <untyped> bigint, got int
```

In other words, you can't put a variable next to a constant when appending to a slice, because the constant is still considered in its untyped version.

### LINKED ISSUE #1341: nil can be converted to invalid types (ie. int)
First discovered in https://github.com/gnolang/gno/pull/1177#issuecomment-1747914727 , I think I have forgotten to make an issue to track it separately.

Sample program:

```
package main

func main() {
        println(int(nil))
}

// Output: 0
```

`int(nil)` makes no sense and should not be accepted.

### PR DESCRIPTION
**Pull request by mvertes**
This is a particular case of Go append builtin where the array type is inferred from the first argument. We detect an untyped const passed as argument, then we convert it to the array type prior to proceed further. It fixes further consistency checks and value generation.

Fixes #1136, fixes #1149, fixes #1341.

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by mvertes** [comment]
Seeing if I can make the fix even more generic and also handle any untyped constant value, as to fix also #1149.

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -1006,6 +1006,23 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
...
+								continue
+							}
+							if t1 := evalStaticTypeOf(store, last, arg); t1 != nil && !isUntyped(t1) {
+								// Consider only untyped values (including nil).
+								continue
+							}
+							// Get the array type from the first argument and convert to it.
+							s0 := evalStaticTypeOf(store, last, n.Args[0])
+							tx := constType(arg, s0.Elem())
+							var arg1 Expr = Call(tx, arg)
```
Comment (thehowl) [inline gnovm/pkg/gnolang/preprocess.go:1023]: Nit:

```suggestion
							arg1 := Call(tx, arg)
```

Even if it returns *CallExpr, it doesn't matter, no?

**Review by thehowl — APPROVED** [review]
LGTM :clap:

**Comment by thehowl** [comment]
> Does that also fix the bug we had on gnochess #1170 ?

Not as far as I can see, alas. This is specifically tackling constant expressions in append; #1170 is about the internal pointers not being fixed with appends.

**Comment by piux2** [comment]
@thehowl, @moul,

Issue #1136 and issue #1149 are rooted in different issues. 

Regarding #1136:
In Go, it's not permissible to append a nil to concrete types like []int. In the test case discussed, error is a built-in interface type. In Go, the default zero value for an interface type is nil, so it's allowable to append a nil to a slice of type []error. However, the behavior described in issue #1149 is different. If one attempts to append a nil to []int, it should, and typically does, cause a panic. Notably, after this fix, this panic no longer occurs when appending a nil to []int.

For #1149:
As the error message hinted. The core of the problem lies in converting variadic generic type arguments into a slice before they are passed into the append function. It's important to note that this issue isn't exclusive to append but also impacts any function that accepts variadic generic type arguments.


We will need to dig more and fix it at the root.

**Comment by thehowl** [comment]
> In Go, it's not permissible to append a nil to concrete types like []int. In the test case discussed, error is a built-in interface type. In Go, the default zero value for an interface type is nil, so it's allowable to append a nil to a slice of type []error. However, the behavior described in issue #1149 is different. If one attempts to append a nil to []int, it should, and typically does, cause a panic. Notably, after this fix, this panic no longer occurs when appending a nil to []int.

I was surprised to find out you're right (because it seemed to me that this PR didn't special-case nil):

```go
package main

func main() {
        var ints []int
        ints = append(ints, nil, nil)
        println(ints)
}

// Output: slice[(0 int),(0 int)]
```

This works correctly. However, this is actually a phenomenon caused by a root issue: we allow converting `nil` to `int` (and presumably other types):

```go
package main

func main() {
        println(int(nil))
}

// Output: 0
```

I think the approach @mvertes has used is correct (given `append([]T, T...)`, attempt to convert the variadic args into T before appending). However, we suffer a different problem about `nil`, thus allowing to do what you mentioned (appending `nil` to a slice of ints).

>As the error message hinted. The core of the problem lies in converting variadic generic type arguments into a slice before they are passed into the append function. It's important to note that this issue isn't exclusive to append but also impacts any function that accepts variadic generic type arguments.

This one, empirically, does not check out (for the last statement). Try this one, and playing out with changing it (from master):

```gopackage main

func main() {
	var n []int
	x := 2
	vari(n, 1, x)
}

func vari(n []int, params ...interface{}) {}
```

It's append which is special in this case.

**Comment by piux2** [comment]
> Does that also fix the bug we had on gnochess #1170 ?
> 
> Note that to reproduce it, you need to broadcast multiple tx, so the current test system in `tests/files` cannot be used.

For issue #1170, the root cause is located in realm.go, the underlying problem arises from the underlying array of the slice not being marked as "dirty" during the persistence of a realm object when that object is assigned to itself. I'm still evaluating the issue to identify an optimal solution.

**Comment by piux2** [comment]
> LGTM, @piux2 can you make a complementary review, please? 

Yes, done

**Comment by piux2** [comment]
I think here is the right place to fix #1149
https://github.com/gnolang/gno/blob/ce258b16cf27d10ac8e62abde2675c3b4ce76b6f/gnovm/pkg/gnolang/types.go#L1163

and add following  condition:



		+ if isUntyped(varg.T) {
	        +   vargt =	defaultTypeOf(varg.T)
		+ }else{
				vargt = varg.T
		 + }

**Comment by piux2** [comment]
We will have further discussion and update the changes

**Review by piux2 — CHANGES_REQUESTED** [review]
we will dig further

**Review by thehowl — APPROVED** [review]

**Comment by zivkovicmilos** [comment]
@piux2 
Ping for unblocking this PR

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -1006,6 +1006,23 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 								n.Args[1] = args1
 							}
 						}
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/preprocess.go:1008]: maybe at least some enclosing else, to differentiate two disjoint cases.

**Comment by jaekwon** [comment]
LGTM

**Comment by thehowl** [comment]
@mvertes I've made a [small commit](https://github.com/gnolang/gno/pull/1177/commits/04614ecea4ba0b02e0ea4d6349c81315b776bde2) to address the comment by Jae. Then I noticed a small chance for optimisation (re-using the result of `Call`), then I also fixed some tests that was failing on the CI :)

If the changes look good to you, please merge. The code's been reviewed by Jae, if there are further changes requested they can be made in a separate PR :)

**Comment by piux2** [comment]
> @piux2 ping

@mvertes sorry, I thought it was already approved 

@thehowl thank you.

**Pull request merged**